### ISSUE TITLE
skip logic error in survey subtest type

### ISSUE THREAD

**Issue by skippskipp** [Tangerine-Community/Tangerine-client#106]
In the larger EGRA instrument being used for testing, there's a survey subtest returning the error message below.  The 'ResultOfMultiple' skip logic also seems to fail in this subtest: Task 4B: Subtraction - Method  
![screenshot_2015-11-24-16-06-52](https://cloud.githubusercontent.com/assets/2474753/11381352/93d9a8a6-92c9-11e5-9b54-c8854b99d9d9.png)

**Comment by chrisekelley** [comment]
I created a test assessment -  "EGMA bug106 Test" - but am unable to reproduce this error on the tablet. I may need more details on this one.

**Comment by rjcorwin** [comment]
Trying to reproduce this bug from code Scott was probably using at the time. Trying commit 860aedd86f2463c3c79df236afffe1b7a3e31cbf from November 16 but no luck. The commit right after that one, https://github.com/Tangerine-Community/Tangerine-client/commit/c198f817c3312393782a1eeda78ea381f8254aa3, fixes a skippable bug so I'm guessing the trouble was somewhere around there.

**Comment by rjcorwin** [comment]
I now see the discrepancy with that subtest between the Tangerine codebase on DatabasesTC and the current Tangerine client code.

## DatabasecTC

<img width="1280" alt="screen shot 2016-01-06 at 7 23 08 am" src="https://cloud.githubusercontent.com/assets/156575/12142528/c0ee4fd8-b446-11e5-82d6-f252c964de40.png">

## Current Tangerine Client code

<img width="719" alt="screen shot 2016-01-06 at 7 23 01 am" src="https://cloud.githubusercontent.com/assets/156575/12142533/cc5b8b9c-b446-11e5-856b-9a112c9cdb62.png">

Chris pointed out the logic for the second question being skipped is located here.
<img width="1185" alt="screen shot 2016-01-06 at 7 26 33 am" src="https://cloud.githubusercontent.com/assets/156575/12142546/f6275c6c-b446-11e5-9a6d-727fb00a292f.png">

@skippskipp I'm unclear as to what that logic means. I'll dig into the code to see what I can find out about this skip logic field.

**Comment by rjcorwin** [comment]
"Progress meter subtitle not showing up" was also something that seemed to be only happening in legacy Assessments.

**Comment by rjcorwin** [comment]
Hotspot -> https://github.com/Tangerine-Community/Tangerine-client/commit/c198f817c3312393782a1eeda78ea381f8254aa3#diff-9e4e0fd076938dc8fd9d086165b209a7R223

**Comment by rjcorwin** [comment]
Getting hotter. This is always returning on that subtest despite renderCount being larger than questions.length https://github.com/Tangerine-Community/Tangerine-client/blame/a702157a7a0487afaf1de429441c085c1390e866/src/js/modules/subtest/prototypes/SurveyRunItemView.coffee#L140

**Comment by rjcorwin** [comment]
In the screenshot, I just clicked on an option in question one that should cause the skip logic for the following question to run. That skip logic never runs because because the line the breakpoint is on results in a `return`. Either the values being compared are incorrect or the logic of the statement is flawed. I think the statement might be flawed because it sure isn't behaving like I would expect. I would expect this to not issue a `return`.
<img width="1280" alt="screen shot 2016-01-06 at 10 25 21 am" src="https://cloud.githubusercontent.com/assets/156575/12146179/4a020f4e-b460-11e5-804d-71459320f171.png">

**Comment by rjcorwin** [comment]
nvm, `@renderCount >= @questions.length` evaluates to `true` which should and does result in the `return`.

**Comment by rjcorwin** [comment]
So the issue must be the values not being what they should be...

**Comment by rjcorwin** [comment]
There seems to be an underlying issue that @renderCount is wrong. This counter increases for every @onQuestionRendered which gets called by Marionette Child Event `render`. The thing is that these renders are happening twice in a particular pattern. Here's a list of the order of the childView.model.id output going through this callback.

```
18e3c546-eeec-67ad-1c11-d1d428d0a31a
415f1dda-81a8-ae66-f77f-428c243fdf61
5bc83aac-5728-7354-9e12-f776c430b0b4
bb30cee7-44c1-89da-ea96-f6605dc98626 
bb30cee7-44c1-89da-ea96-f6605dc98626
5bc83aac-5728-7354-9e12-f776c430b0b4
415f1dda-81a8-ae66-f77f-428c243fdf61
18e3c546-eeec-67ad-1c11-d1d428d0a31a
```

Definitely a pattern. I goes start to finish and then back again.

**Comment by rjcorwin** [comment]
This may be the root of the problem, maybe not. Either way this has to be sorted out. Either this is happening this way because of Marionette (or how we're using it) or because those Views are rendering themselves again... or duplicating themselves given that the childView.cid popping out is different every time, not following that forward and backward pattern. That really points towards Marionette creating two sets of child views, the second set which is never shown... Maybe WE are calling render on this View when the Marionette is calling it via Layouts...

**Comment by chrisekelley** [comment]
onQuestionRendered() is triggered by the childEvent rendered. I don't trust it. We need to see what is triggering rendered. I was expecting that "add:child" would help out, but I believe that event is bubbled up to AssessmentCompositeView. Anyway, I think there may be a better event to use.

**Comment by rjcorwin** [comment]
I'm getting myself pretty lost in this one, having trouble making heads or tails of things. Something that does seem fishy is the fact that QuestionRunItemView triggers "answer" in all sorts of situations (intentional?) like when a button in the question is done rendering, when the Question is rendered, when it changes... The only thing that makes sense to me for an `answer` event would be when there is user input indicating an answer. 

https://github.com/Tangerine-Community/Tangerine-client/blob/9bbf4461ddb14a2bab4ed3ea3e8fcca2da09bb52/src/js/modules/question/QuestionRunItemView.coffee#L75

**Comment by rjcorwin** [comment]
SurveyRunItemView.render() is definitely getting called twice. But how?

<img width="1279" alt="screen shot 2016-01-06 at 4 43 17 pm" src="https://cloud.githubusercontent.com/assets/156575/12155752/e865a2ca-b494-11e5-8dcf-91c10e708702.png">

Looking at the callstack for the first call...

<img width="711" alt="screen shot 2016-01-06 at 4 46 04 pm" src="https://cloud.githubusercontent.com/assets/156575/12155775/0405e288-b495-11e5-92b7-5134a1adbe4a.png">

Looking at the callstack for the second...

<img width="710" alt="screen shot 2016-01-06 at 4 46 48 pm" src="https://cloud.githubusercontent.com/assets/156575/12155797/2543561a-b495-11e5-98a8-4bc7200db1c4.png">

**Comment by rjcorwin** [comment]
Is it a result of the that `resortView` reference?

**Comment by rjcorwin** [comment]
@chrisekelley Tapping out for the day on this one.

**Comment by chrisekelley** [comment]
This is a nice list of view triggers (that I forgot to hit submit on last night): https://marionette.gitbooks.io/marionette-guides/content/en/views/triggers.html

**Comment by chrisekelley** [comment]
When the form renders, the second question, "What other methods...", should be hidden. Before everything is rendered to the screen, the app should check if any of the questions should be hidden by running its skipLogicCode, if available. The method updateSkipLogic() checks for this. If the question should be skipped, it adds "disabled_skipped" to that questionView 

``` javascript
qv.$el.addClass("disabled_skipped");
```

I am thinking that either the wrong $el is getting hidden, or the app is not performing updateSkipLogic() at the right time.

**Comment by chrisekelley** [comment]
You are correct about the double-rendering. The first time it renders, the second question, "What other methods..." is correctly rendered as hidden. The second time, no. I am seeing this when I set a breakpoint at AssessmentCompostieView.addChildPostRender() and peek into currentSubtest.questionViews[1].$el. So my question now is why updateSkipLogic() is not getting hit the second time?

**Comment by chrisekelley** [comment]
Trying to track down why updateSkipLogic()  is not getting executed the second time. The second time the whole view is rendered, onQuestionRendered() is not being triggered by a 'render' event.

**Comment by chrisekelley** [comment]
Also, childview:rendered event triggers onQuestionRendered()

**Comment by chrisekelley** [comment]
OK, I think I have a solution. I noticed that on the second render, onRenderCollection() is triggered. That method triggers 'subRendered', which can be caught by AssessmentCompositeView. I added that event to ACV's childEvents, and associated it with a new method "subRendered()': 

``` javascript
  subRendered: ->
    console.log("subRendered")
    currentSubtest = @children.findByIndex(0)
    currentSubtest.updateSkipLogic()
```

This is similar to what I did in addChildPostRender() for focusMode.

**Comment by chrisekelley** [comment]
My solution fails on the large LTTP2 Assessment. As it loops through each questionView,, SurveyRunItemView should execute onQuestionRendered(). It executes it for the first question, but not the second, which is the one with the skipLogicCode. Unlike the smaller test, it also does not execute onRenderCollection().

**Comment by rjcorwin** [comment]
[1/8/16, 12:05:11 PM] Scott Kipp: it looks like some subtest-level skip logic may not be working.  this piece isn't functioning:   @skip() if ResultOfGrid("add") < 1
[1/8/16, 12:05:45 PM] Scott Kipp: it's in the 'Action on Display' field of this subtest:  http://databases.tangerinecentral.org/group-rti_tanzania_2016/_design/ojai/index.html#subtest/d68e1de9-cb25-7020-25a3-dce305e41cbd
[1/8/16, 12:06:22 PM] Scott Kipp: basically if there are no correct answers in the previous grid (no values of "1" for any variable), that subtest should be skipped

**Comment by chrisekelley** [comment]
I tested this on the TC server version - i marked all of the answers on EGMA 4 wrong, and it skipped to EGMA 5.

**Comment by rjsteinert** [comment]
Fixed with PR #209
